### PR TITLE
[fixing cuda launch config failure on UpSampleNearest]

### PR DESCRIPTION
**Pull request by jjsjann123**
Adding limitation on launch config for grid size 
Test added in test_cuda;

### CODE REVIEW

File: aten/src/ATen/native/cuda/UpSampleNearest1d.cu [hunk 1]
```diff
@@ -109,6 +109,15 @@ static void upsample_nearest1d_out_cuda_template(
...
   dim3 gdim{cuda::ATenCeilDiv(n, bdim.x)};
+  // Error out on cases where gdim.x exceeds limit of launch config, as
+  // the current kernel implementation doesn't loop over non-batch dimensions.
+  // Note that it is unlikely to happen, as maxGridSize[0] is 2**31-1
+  // TODO: kernel implementation could stride on spatial dimension. We probably
+  //       need to overhaul the kernel.
+  TORCH_CHECK(
+      gdim.x <= at::cuda::getCurrentDeviceProperties()->maxGridSize[0],
+      "input tensor has spatial dimension larger than the kernel capacity");
+
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/UpSampleNearest1d.cu]: tensors are indexed with ints in the kernel, so just check that the number of elements fits, and don't do this check.

File: aten/src/ATen/native/cuda/UpSampleNearest1d.cu [hunk 2]
```diff
@@ -162,6 +171,15 @@ static void upsample_nearest1d_backward_out_cuda_template(
...
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/UpSampleNearest1d.cu]: same here

File: test/test_nn.py [hunk 3]
```diff
@@ -8391,6 +8391,41 @@ def test_rnn_retain_variables(self, device, dtype):
             with torch.backends.cudnn.flags(enabled=False):
                 self._test_rnn_retain_variables(device, dtype)
 
+    @onlyCUDA
```
Comment (ngimel) [inline test/test_nn.py]: How much memory do those tests require? Regular CI machines are M60s with 6GB memory IIRC, so if you need more, use large memory decorator.

**Review by ngimel — APPROVED** [review]
Looks good, just a couple small fixes.

File: test/test_nn.py (hunk 3 above)
Comment (jjsjann123) [reply test/test_nn.py]: Looking at the size of input + output, it takes 0.625GB. Given the way input/output are allocated in the code path, it should have cached 0.75GB with 0.625GB actually allocated.

Running the test independently I'm seeing 2.5GB cached memory in total at the end of the test, so we should be safe here.

File: aten/src/ATen/native/cuda/UpSampleNearest1d.cu (hunk 1 above)
Comment (jjsjann123) [reply aten/src/ATen/native/cuda/UpSampleNearest1d.cu]: It's fair to assume that maxGridSize is not going to change in the future.
I'll update it for the places where it's the same kernel launch.

**Comment by jjsjann123** [comment]
Took me a while wondering how that expected failure is passing the test, then realized it's rocm...
I'll just exclude rocm from that, since I have no idea what are the limitations. Feel free to add a rocm specific failure test.

File: test/test_nn.py [hunk 4]
```diff
@@ -8417,8 +8417,13 @@ def test_upsamplingNearest3d_launch_config(self, device):
         inp_ref = inp.cpu()
         out_ref = m(inp_ref)
         self.assertEqual(out_ref, out)
+        print(torch.cuda.memory_cached())
```
Comment (ngimel) [inline test/test_nn.py]: please remove prints and sleep

**Comment by jjsjann123** [comment]
Can't get any idea out of failing CI. Should I be concerned?

**Comment by ngimel** [comment]
Failures are unrelated

**Comment by zhanwenchen** [comment]
When is this getting released? Right now I had to implement a Dataloader for inference.

**Comment by ngimel** [comment]
It is in nightly releases.

**Pull request closed without merging**